chore(tests): adds utilities for mocking Elgg data, testing queries

hypeJunction opened the pull request:
Adds a set of utilities for mocking entities, metadata, annotations and
relationships.
Adds utilities for testing database queries.

hypeJunction commented:
@mrclay Any idea why update and delete subtype tests are failing here? I ran out of debugging ideas.

mrclay commented on engine/tests/phpunit/Elgg/Database/SubtypeTableTest.php:
```diff
@@ -0,0 +1,142 @@
+				'id' => 1,
+				'type' => 'object',
+				'subtype' => 'foo',
+				'class' => '\FooObject',
```
Why the leading `\`? This could cause problems.

hypeJunction commented:
What's your general impression from this? I feel that the complexity is worth the gain. We can break down and refactor later.

mrclay commented on engine/tests/phpunit/Elgg/TestCaseMocks.php:
```diff
@@ -0,0 +1,1042 @@
+	 * @param array  $results An array of results
+	 * @return \stdClass|null
+	 */
+	public function fetchFromQueryResult($hash, $results) {
```
I'm tracking down a bug in this function

hypeJunction replied:
Yes, I had an aha moment a few minutes ago. Working on it

hypeJunction replied:
Or are you making a PR?

hypeJunction replied:
We can loose this now? Or?

mrclay replied:
Yeah, fetchFromQueryResult can go.

mrclay commented:
one was that `setDatabaseQueryResult` needed a way to set the rowCount return value for update queries. I've added that. We also need similar for lastInsertID(); a simple counter isn't going to be sufficient when we start using this.

hypeJunction commented:
I am leaving this to you for now. Realized how tired I am. Good night.

hypeJunction commented on engine/tests/phpunit/Elgg/TestCaseMocks.php:
```diff
@@ -0,0 +1,1070 @@
+		$statement->expects($this->test->any())
+				->method('fetch')
+				->will($this->test->returnCallback(function() use (&$results) {
+							return array_shift($results);
```
Wow!

hypeJunction commented:
Test are green, just the commit message. I will squash and clean up jnline docs in the class a bit. 
I want to get in a sample test for an action so we have a benchmark. I think I will use the register action and we can test drive the entity mocks here.

mrclay commented:
Definitely seems helpful in some spots. The bad is, of course, it's very easy to build a faulty component that passes tests by misunderstanding SQL. Hitting a real DB is so much more reliable and our Simpletest suite has been a lifesaver in my refactoring work.

I'll try to simplify something like the PersistentLoginTest mess to get more hands on with it.

hypeJunction commented:
@mrclay I had a change of hearts. I think this new approach will help us with test coverage (with mock builder we just skip a bunch of code that we could be analyzing).

hypeJunction commented:
Icons tests are failing for whatever reason. I don't have any more time left, so will have to finish it some time next week.

mrclay commented on engine/classes/Elgg/Database/AccessCollections.php:
```diff
@@ -45,38 +109,38 @@ public function __construct($site_guid) {
+
 		if ($user_guid == 0) {
-			$user_guid = _elgg_services()->session->getLoggedInUserGuid();
+			$user_guid = $this->session->getLoggedInUserGuid();
```
This is improvement but the test (`ElggCoreAccessSQLTest::testLoggedOutUser`) is not prepared. It only changes the session in the service provider. It would need to inject the new session into this instance, or rebuild the instance. I'll work on it.

hypeJunction replied:
Thanks. I fixed the icon test (subtype resolution wasn't done right), so it's just the simple tests that are failing. Scrutinizer says it's a 3% increase in test coverage. I am drained, need to clear my head and rest my body.

mrclay commented:
Sent you a PR to finish this.

hypeJunction commented:
@mrclay, I have added `ElggMetadata` tests and some metastrings mocking logic

hypeJunction commented:
Added some ElggObject tests. Relationship table mock now uses query specs.

mrclay commented on engine/classes/Elgg/Database/MetastringsTable.php:
```diff
@@ -112,8 +112,11 @@ private function getIdCaseSensitive($string) {
-			$query = "SELECT id FROM {$this->getTableName()} WHERE string = BINARY '$escaped_string' LIMIT 1";
-			$results = $this->db->getData($query);
+			$query = "SELECT id FROM {$this->getTableName()} WHERE string = BINARY :string LIMIT 1";
+			$params = [
```
When converting to params, eliminate the sanitize call. FWIW I prefer the other syntax for up to a couple params:
```php
$query = "SELECT id FROM {$this->getTableName()} WHERE string = BINARY ? LIMIT 1";
$results = $this->db->getData($query, null, [$string]);
```

hypeJunction replied:
It's easier to use params in tests.

mrclay commented on engine/classes/Elgg/Database/MetastringsTable.php:
```diff
@@ -157,7 +160,12 @@ private function getIdCaseInsensitive($string) {
 	function add($string) {
 		$escaped_string = $this->db->sanitizeString(trim($string));
```
This goes away...

mrclay commented on engine/classes/ElggObject.php:
```diff
@@ -143,24 +152,31 @@ protected function create() {
 		
-		$guid = (int)$this->guid;
+		$guid = (int) $this->guid;
 		$title = sanitize_string($this->title);
```
Remove these...

hypeJunction commented:
FYI, the PR increases test coverage by 8%. https://scrutinizer-ci.com/g/Elgg/Elgg/inspections/4376360f-bbfc-4ebb-9dc4-4949c130c98c

mrclay commented on engine/classes/Elgg/Database/EntityTable.php:
```diff
@@ -14,48 +15,120 @@
+	 * @param \stdClass $row Entity base information
+	 * @return int|false
+	 */
+	public function insertRow(\stdClass $row) {
```
I'm OK with this, but we should start exposing the built-in DBAL connection methods to internal API. It already has this kinda stuff.

hypeJunction replied:
I needed this method for mocking.

hypeJunction replied:
Could you elaborate what you meant by your comment?

mrclay replied:
Internally we use a Doctrine DBAL connection. It has some shortcuts that would be useful: http://docs.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/data-retrieval-and-manipulation.html#delete

hypeJunction replied:
Ok. I will keep in mind during future refactor.  For now, I just want to merge this and get back to some down to earth work.

mrclay commented on engine/tests/phpunit/Elgg/Mocks/Database/MetastringsTable.php:
```diff
@@ -0,0 +1,53 @@
+	/**
+	 * {@inheritdoc}
+	 */
+	public function __construct(\Elgg\Cache\Pool $cache, \Elgg\Database $db) {
```
Doesn't seem needed

hypeJunction replied:
What's not needed?

hypeJunction replied:
Ah, you mean the constructor. I am keeping them in for now, until we update all `private` properties to `protected`.

mrclay commented on engine/tests/phpunit/Elgg/Mocks/Database/MetastringsTable.php:
```diff
@@ -0,0 +1,53 @@
+	 * {@inheritdoc}
+	 */
+	public function add($string) {
+		$escaped_string = $this->db->sanitizeString(trim($string));
```
Don't escape if using params

mrclay commented:
What are these files: `guid][]` and `query_specs[-`?

hypeJunction commented:
Hmm. Looks like I was typing something during rebase. I will investigate tot

hypeJunction commented on engine/classes/Elgg/Database.php:
```diff
@@ -276,7 +276,7 @@ public function updateData($query, $get_num_rows = false, array $params = []) {
 			return $stmt->rowCount();
 		} else {
-			return true;
+			return (bool) $stmt->rowCount();
```
@mrclay This function was lying about the result. Does this look ok to you?

hypeJunction replied:
This was a bad assumption. I will instead throw from database mock for non-select queries if no spec is defined, so that we don't miss any queries we should be mocking.

hypeJunction commented:
Added Datalist tests and expanded Metastrings tests.
Query specs now takes a callable for results instead of an array.

I am going to stop now, so we can finalize and merge - it's getting too big.

I am confident we can get our test coverage up to over 50% by 2.3.

hypeJunction committed:
chore(tests): adds a PHPUnit mocking service

Adds a service for mocking Elgg entities, metadata, annotations,
and relationships.
Adds utitilites for testing database queries.

mrclay commented on engine/tests/phpunit/Elgg/Mocks/Database.php:
```diff
@@ -0,0 +1,224 @@
+			if ($spec->sql == $sql && $params == $spec->params) {
+				$match = true;
+				if (isset($spec->results) && is_callable($spec->results)) {
+					$results = call_user_func($spec->results, $spec);
```
If it's an array but not callable, just assume it's the results array.

mrclay commented on engine/classes/ElggObject.php:
```diff
@@ -143,24 +150,27 @@ protected function create() {
-		return $this->getDatabase()->updateData($query) !== false;
+		$params = [
+			':guid' => $this->guid,
+			':title' => $this->title,
```
cast string.

mrclay commented:
I bravely suggest we merge and I'll try to merge into master tonight.

hypeJunction commented:
Ok. Give me a few, addressing your comments. Will merge once all green.

hypeJunction commented:
Making a merge to master now.

hypeJunction commented:
@Elgg/contributors No more excuses for not including tests with pull requests :stuck_out_tongue:
